near-sdk/src/store/index_map.rs

```diff
@@ -0,0 +1,152 @@
+use std::fmt;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use once_cell::unsync::OnceCell;
+
+use crate::utils::StableMap;
+use crate::{env, CacheEntry, EntryState, IntoStorageKey};
+
+const ERR_ELEMENT_DESERIALIZATION: &str = "Cannot deserialize element";
+const ERR_ELEMENT_SERIALIZATION: &str = "Cannot serialize element";
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub(crate) struct IndexMap<T>
+where
+    T: BorshSerialize,
+{
+    pub(crate) prefix: Box<[u8]>,
+    /// Cache for loads and intermediate changes to the underlying index map.
+    /// The cached entries are wrapped in a [`Box`] to avoid existing pointers from being
+    /// invalidated.
+    #[borsh_skip]
+    pub(crate) cache: StableMap<u32, OnceCell<CacheEntry<T>>>,
```

Comment:
I think its best to also mention why we're using `u32` in the comments somewhere over `usize`. I remember it being addressed in one of the older PRs but don't know where it was placed